server/base/src/main/java/org/apache/accumulo/server/conf/codec/VersionedProperties.java

```diff
@@ -106,10 +107,15 @@ public VersionedProperties(final int dataVersion, final Instant timestamp,
    * value should be used on data writes as the expected version. If the data write fails do to an
    * unexpected version, it signals that the node version has changed since the instance was
    * instantiated and encoded.
+   * <p>
+   * Implementation note: The data version is stored and returned is an unsigned 32-bit integer
+   * value. Internally, ZooKeeper stores the value as a 32-bit signed value that can roll-over and
+   * become negative. The can break applications that rely on the value to always increase. This
+   * class ensures that the long returned will increase and will not be negative.
```

Comment:
What applications would rely on this always increasing?

Doing this will make it never be negative, but it could still roll over and start back at zero, so it's not guaranteed to always increase. Right?

Reply:
Second part first - yes, it will potentially rollover back to zero.  To prevent that would require an upstream change to ZooKeeper, but 2^32 versions should last us a bit.

The update count in ZooBasedConfiguration expects that it will always increase and with the recent changes that count is the data version count. Specifically the code at https://github.com/apache/accumulo/blob/a30980f8a99e3eef2aa828f591e8443818478c87/server/base/src/main/java/org/apache/accumulo/server/conf/ZooBasedConfiguration.java#L84 where the count is calculated with ` count += dataVersion`

Reply:
> Second part first - yes, it will potentially rollover back to zero.

In that case, the javadoc wording is suspect.
```suggestion
   * become negative. The can break applications that rely on the value to always increase. This
   * class avoids rollover by doubling the positive values and ensures it will never be negative.
```

> The update count in ZooBasedConfiguration expects that it will always increase

That might be okay for now... but it would be nice to change that mechanism so it can detect changes in the hierarchy without caring about the numeric value.

Reply:
Yes it would - but that would require changes to `AccumuloConfiguration` and  sub-classes like `ConfigurationCopy`, which I thought was agreed that would be best left for 3.0

Reply:
> Yes it would - but that would require changes to `AccumuloConfiguration` and sub-classes like `ConfigurationCopy`, which I thought was agreed that would be best left for 3.0

It would be nice to fix for 2.1 if it's not too much trouble, but it depends on how big of a change it would be. My guess (without looking at it at all) is that it wouldn't be too bad, but I have a habit of biting off more than I can chew. :smiley_cat: